src/pid/maps.rs

```diff
@@ -0,0 +1,243 @@
+//! Process memory mappings information from `/proc/[pid]/maps`.
+
+use std::ffi::OsString;
+use std::io::{self, BufRead};
+use std::os::unix::ffi::OsStringExt;
+use std::path::PathBuf;
+use std::{fs, ops};
+
+use libc;
+use nom::{rest, space};
+
+use parsers::{map_result, parse_usize_hex, parse_u32_hex, parse_u64, parse_u64_hex};
+use unmangle::unmangled_path;
+
+/// Process memory mapping information.
+///
+/// Due to the way paths are encoded by the kernel before exposing them in
+/// `/proc/[pid]/maps`, the parsing of `path` and `is_deleted` is
+/// ambiguous. For example, all the following path/deleted combinations:
+///
+/// - `/tmp/a\nfile` *(deleted file)*
+/// - `/tmp/a\nfile (deleted)` *(existing file)*
+/// - `/tmp/a\\012file` *(deleted file)*
+/// - `/tmp/a\\012file (deleted)` *(existing file)*
+///
+/// will be mangled by the kernel and decoded by this module as:
+///
+/// ```rust,ignore
+/// MemoryMapping (
+///    ...,
+///    path: PathBuf::from("/tmp/a\nfile"),
+///    is_deleted:true,
+/// )
+/// ```
+///
+/// If the `path` of a mapping is required for other than purely informational
+/// uses (such as opening and/or memory mapping it), a more reliable source
+/// (such as `/proc/[pid]/map_files`) should be used, if available. The open
+/// file `(dev, inode)` should also be checked against the values provided by
+/// the mapping.
+///
+/// See `man 5 proc`, `Linux/fs/proc/task_mmu.c`, and `Linux/fs/seq_file.c`.
+#[derive(Debug, PartialEq, Eq, Hash)]
+pub struct MemoryMapping {
+    /// Address range that this mapping occupies in the process virtual memory
+    /// space.
+    pub range: ops::Range<usize>,
+    /// Whether pages in this mapping may be read.
+    pub is_readable: bool,
+    /// Whether pages in this mapping may be written.
+    pub is_writable: bool,
+    /// Whether pages in this mapping may be executed.
+    pub is_executable: bool,
+    /// Whether this mapping is shared.
+    pub is_shared: bool,
+    /// Offset into the file backing this mapping (for non-anonymous mappings).
+    pub offset: u64,
+    /// Device containing the file backing this mapping (for non-anonymous
+    /// mappings).
+    pub dev: libc::dev_t,
+    /// Inode of the file backing this mapping (for non-anonymous mappings).
+    pub inode: libc::ino_t,
+    /// Path to the file backing this mapping (for non-anonymous mappings),
+    /// pseudo-path (such as `[stack]`, `[heap]`, or `[vdso]`) for some special
+    /// anonymous mappings, or empty path for other anonymous mappings.
+    pub path: PathBuf,
```

Comment:
I think it would be better to parse out the map type into a proper enum, I think putting invalid paths into a `PathBuf` is a no-no.  Maybe could look something like this:

```
pub enum MapType {
  Anonymous,
  File(pub PathBuf),
  Stack,
  Heap,
  Vdso,
  Vvar,
  Vsyscall,
}
```

Reply:
> However, enumerating all possible pseudo-paths might be problematic, as the exact list is not documented, is architecture and config dependent, and might change (new pseudo-paths might be added) in future kernel versions (which would break this API).

Yeah, it would be nice if the `#[non_exhaustive]` attribute was stable.  How about adding all of the known/common variants and compromising on an `Unknown(String)` variant as a fallback for everything else?  (I think `String` is OK here, since the file is guaranteed to be valid ascii, right?)